File: undirected_motifs.py
# libraries
    
import numpy as np
from scipy.stats import norm
from scipy.sparse import csr_matrix
import networkx as nx

# inputs
# n = number of vertices
# p = probability of an edge being included
# p^2(1 + a_conn) = probability of the connected motif
# p^2(1 + a_disj) = probability of the disjoint motif (this is technically
# not a motif but is included to complete the coherent configuration)

# global variables
# R_conn = indicator matrix for connected motif

# parameters: n (int), disp (optional boolean)
# returns: list of integer matrices [rho_id, rho_conn, rho_disj]; matrices computed for n
# optional: if disp=True (default disp=False), displays rho_id, rho_conn, and rho_disj

def undirected_rho_list(n, disp=False):

  # rho matrices
  rho_id = np.identity(3).astype(int)

  rho_conn = np.array([[0,1,0],
                       [2*n-4,n-2,4],
                       [0,n-3,2*n-8]]).astype(int)

  rho_disj = np.array([[0,0,1],
                       [0,n-3,2*n-8],
                       [(n-2)*(n-3)/2,(n-3)*(n-4)/2,(n-4)*(n-5)/2]]).astype(int)

  # display rho matrices if disp=True
  if disp:
    print(\"rho_id \\n{} \\n \\nrho_conn \\n{} \\n \\nrho_disj \\n{}\".format(rho_id, rho_conn, rho_disj))

  # returns list of rho matrices
  return [rho_id, rho_conn, rho_disj]

# example code

undirected_rho_list(6,disp=True)
          
# parameters: n (int), a_conn (float), a_disj (float)
# returns: if square root exists, list of three floats b; coefficients
# of positive definite square root, otherwise "Square root does not exist"

def square_root_coefficients(n, a_conn, a_disj):

  # a_id = 1 normalization
  a_id = 1

  # lists of coefficients and rho matrices
  a = [a_id, a_conn, a_disj]
  rho_list = undirected_rho_list(n)
          
  # linear combination
  rho = np.dot(a, rho_list)\n",

  # positive definite square root; returns "Square root does not exist" if
  # square root does not exist
  d, p = np.linalg.eig(rho)
  if sum(d < 0) > 0:
    return "Square root does not exist"
  d_root = np.diag(np.sqrt(d))
  rho_root = p@d_root@np.linalg.inv(p)

  # constructs gram matrix and inhomogeneous part
  gram = np.zeros((3,3))
  for i in range(3):
    for j in range(3):
      gram[i,j] = np.trace(np.transpose(rho_list[i])@rho_list[j])

  vec = np.zeros(3)
  for i in range(3):
    vec[i] = np.trace(np.transpose(rho_root)@rho_list[i])

  # solves linear system
  b = np.linalg.solve(gram, vec)

  # returns coefficients
  return b
          
# example code

print(square_root_coefficients(10, 0, 0))
print(square_root_coefficients(10, 0.1, 0))
print(square_root_coefficients(10, 1, 0))
          
# implements edge ordering e_0 = 1 <-> 0, e_1 = 2 <-> 0, e_2 = 2 <-> 1,
# e_3 = 3 <-> 0, e_4 = 3 <-> 1, ..., e_{n*(n-1)/2-1} = n-1 <-> n-2
# edge e_k = i <-> j where k = max*(max-1)/2 + min with max = max(i,j) and min = min(i,j)

# parameters: i (int), j (int)
# returns: k (int); labeled undirected edge connecting nodes i and j

def edge(i,j):
  if i > j:
    return int(i*(i-1)/2 + j)\n",
  else:
    return int(j*(j-1)/2 + i)\n",

# parameters: k (int)
# returns: i (int), j (int); labeled nodes connected by undirected edge k with i > j

def edge_inv(k):
  i = 1
  while (i+1)*i/2 <= k:
    i += 1
  j = int(k - i*(i-1)/2)
  return i, j
          
# parameters: n (int)
# returns: nothing; constructs glogal variable R_conn

def undirected_row_col_conn(n):

# initialize arrays for length
data = row = col = np.ones(n*(n-1)*(n-2)).astype(int)
index = 0

for i in range(0,n):
  for j in range(0,i):
    for k in range(0,j):

    # three undirected edges connecting three nodes
    e_1 = edge(i,j)\n",
    e_2 = edge(j,k)\n",
    e_3 = edge(k,i)\n",

    # non-zero entries for connected motif
    row[index] = e_1
    col[index] = e_2
    index += 1

    row[index] = e_2
    col[index] = e_1
    index += 1

    row[index] = e_2
    col[index] = e_3
    index += 1

    row[index] = e_3
    col[index] = e_2
    index += 1

    row[index] = e_3
    col[index] = e_1
    index += 1

    row[index] = e_1
    col[index] = e_3
    index += 1

    # m = number of possible undirected edges
    m = int(n*(n-1)/2)

    # global variable R_conn
    global R_conn

    # constructs sparse matrix R_conn
    R_conn = csr_matrix((data, (row, col)), shape = (m, m)).toarray()

    # returns nothing
    return  

# example code
  
undirected_row_col_conn(6)

print(R_conn)
          
# parameters: n (int), p (float), a_conn (float), a_disj (float)
# returns: edges for random graph
# note: n must match n used for global variables row and col from undirected_row_col_conn

def undirected_random_graph(n, p, a_conn, a_disj, disp_iid=False, disp_corr=False):

  # m = number of possible undirected edges
  m = int(n*(n-1)/2)

  # vector of independent standard normal random variables for each possible edge
  x_iid = np.random.normal(0,1,m)

  # displays x_iid if disp_iid=True
  if disp_iid:
    print(\"x_iid \\n{}\".format(x_iid))

  # computes coefficients of positive definite square root
  [b_id, b_conn, b_disj] = square_root_coefficients(n, a_conn, a_disj)

  # vector of correlated standard normal random variables for each possible edge
  x_corr = (b_id-b_disj)*x_iid + (b_conn-b_disj)*R_conn@x_iid + b_disj*np.sum(x_iid)*np.ones(m)

  # displays x_corr if disp_corr=True
  if disp_corr:
    print(\"x_corr \\n{}\".format(x_corr))

  # t = threshold value for including edge
  t = norm.ppf(p)

  # includes edges based on x_corr and the threshold t
  edges = [1 if element < t else 0 for element in x_corr]

  # returns edges
  return edges
  
# example code
  
undirected_row_col_conn(6)

print(R_conn)

undirected_random_graph(6, 0.1, 0, 0, disp_iid=True, disp_corr=True)
          
# parameters: n (int), edges (int list)
# returns: adjacency matrix for edges

def undirected_adj(n, edges):

  adj = np.zeros((n,n)).astype(int)

  k = 0
  while k < len(edges):
    if edges[k] == 1:
      i, j = edge_inv(k)
      adj[i,j] = adj[j,i] = 1
  k += 1

  return adj
          
# example code
# optional: a_disj = -4*a_conn/(n-3) to make 2*(n-2)*a_conn + ((n-2)*(n-3)/2)*a_disj = 0

n = 10

# create global variable R_conn for given n

undirected_row_col_conn(n)

# for given n generate random graphs for any value of p and any values of
# a_conn and a_disj for which square root exists

edges = undirected_random_graph(n,0.25,0,0)
adj = undirected_adj(n, edges)
graph = nx.from_numpy_matrix(adj)

print(\"edges \\n{} \\n \\nadj \\n{} \\n\".format(edges, adj))
nx.draw(graph)
      
# example code
# optional: a_disj = -4*a_conn/(n-3) to make 2*(n-2)*a_conn + ((n-2)*(n-3)/2)*a_disj = 0

n = 100

# create global variable R_conn for given n

undirected_row_col_conn(n)

# for given n generate random graphs for any value of p and any values of
# a_conn and a_disj for which square root exists

edges = undirected_random_graph(n,0.1,0,0)
adj = undirected_adj(n, edges)
graph = nx.from_numpy_matrix(adj)

nx.draw(graph)
      
# example code

edges = undirected_random_graph(n,0.1,0.00001,0)
adj = undirected_adj(n, edges)
graph = nx.from_numpy_matrix(adj)

nx.draw(graph)
      
# example code

edges = undirected_random_graph(n,0.05,0,0)
adj = undirected_adj(n, edges)
graph = nx.from_numpy_matrix(adj)

nx.draw(graph)
      
# example code

edges = undirected_random_graph(n,0.05,0.00001,0)
adj = undirected_adj(n, edges)
graph = nx.from_numpy_matrix(adj)

nx.draw(graph)

File: undirected_motifs.java
/*
    This is a Java implementation of the algorithm published in the article
    "Motifs, coherent configurations and second order network generation"
    in the journal Physica D: Nonlinear Phenomena by Jared Bronski and Tim Ferguson.
    The algorithm generalizes Erdos-Renyi random graphs be introducing correlation between
    the random variables used to determine if an edge is included or not. This algorithm is
    based on original insights into the structure of these graphs and uses an algebraic
    object called a coherent configuration to exactly replicate the desired distribution with
    high efficiency.
    
    Complexity analysis (let n denote the number of vertices in the random graph):
    The algorithm requires O(n^2) storage which is best possible since graphs from the desired distribution are dense (i.e. have order n^2 edges).
    The algorithm requries O(n^3) calculations which is the same complexity as the motifs of a graph with n vertices. Note that these calculations occur
    in three main stages the first of which is O(1) while the second two are O(n^2) and O(n^3) but can be completed with a parallel architecture.

    Journal publication: https://www.sciencedirect.com/science/article/abs/pii/S0167278921002657
    arXiv preprint: https://arxiv.org/abs/1808.05076
    Medium post: https://medium.com/@timferguson000/efficient-generation-of-erd%C3%B6s-r%C3%A9nyi-random-graphs-with-prescribed-motif-distributions-bd4581134447
 */

package undirectedmotifs;

import java.lang.Math;
import Jama.*; // documentation https://math.nist.gov/javanumerics/jama/doc/Jama/Matrix.html
import java.util.Random; 
import org.jgrapht.*; // documentation https://jgrapht.org/guide/UserOverview#hello-jgrapht
import org.jgrapht.graph.*;

public class UndirectedMotifs {
    
    /*
       Given parameter n:
       Construct rho matrices
    */
    
    public static Matrix[] rhoMatrices(int n) {
        
        double[][] rho_0 = {
                            {1,0,0},
                            {0,1,0},
                            {0,0,1}
                           };       
        double[][] rho_1 = {
                            {0,1,0},
                            {2*n-4,n-2,4},
                            {0,n-3,2*n-8}
                           };
        double[][] rho_2 = {
                            {0,0,1},
                            {0,n-3,2*n-8},
                            {(n-2)*(n-3)/2,(n-3)*(n-4)/2,(n-4)*(n-5)/2}
                           };
        
        return new Matrix[]{
                            new Matrix(rho_0,3,3),
                            new Matrix(rho_1,3,3),
                            new Matrix(rho_2,3,3)
                           };
    }
    
    /*
       Given parameters n, a_0, a_1, a_2:
       Determine if the covariance matrix is positive definite
    
       If covariance matrix is positive definite:
       Compute coefficients of the positive definite square root of rho
       Set indicator to 1.0
       Print that parameters are admissible and print parameter values
    
       If covariance matrix is no positive definite:
       Keep default indicator value 0.0
       Print that parameters are not admissible and print parameter values
       
    */
    
    public static double[] coefficientsSqrt(int n, double a_0, double a_1, double a_2) {
        
        /*
           Default values of coefficient and indicator variables
        */
        
        double b_0 = 0.0;
        double b_1 = 0.0;
        double b_2 = 0.0;
        double indicator = 0.0;
        
        /*
           Constructs matrix part of matrix equation
        */
        
        Matrix[] rho_matrices = rhoMatrices(n);
        double[][] a = new double[3][3];
        for (int i = 0; i < 3; i++) {
            for (int j = 0; j < 3; j++) {
                a[i][j] = ((rho_matrices[i].transpose()).times(rho_matrices[j])).trace();
            }
        }
        Matrix A = new Matrix(a,3,3);

        Matrix rho = (rho_matrices[0].times(a_0)).plusEquals(rho_matrices[1].times(a_1)).plusEquals(rho_matrices[2].times(a_2));  
        Matrix D = rho.eig().getD();
        double d_0 = D.get(0,0);
        double d_1 = D.get(1,1);
        double d_2 = D.get(2,2);
        
        /*
           Test if covariance matrix is positive definite by checking if eigenvalues are positive
        */
        
        if ((d_0 > 0) && (d_1 > 0) && (d_2 > 0)) {
            
            indicator = 1.0; // Sets indicator to 1.0
            
            double d_0_sqrt = Math.sqrt(d_0);
            double d_1_sqrt = Math.sqrt(d_1);
            double d_2_sqrt = Math.sqrt(d_2);
            Matrix D_sqrt = new Matrix(new double[][]{
                                                      {d_0_sqrt,0,0},
                                                      {0,d_1_sqrt,0},
                                                      {0,0,d_2_sqrt}
                                                     },3,3); // Computes square root of diagonal eigenvalue matrix
            Matrix V = rho.eig().getV();
            Matrix rho_sqrt = V.times(D_sqrt.times(V.transpose())); // Constructs positive definite square root of rho
            double[] y = new double[3];
            for (int i = 0; i < 3; i++) {
                y[i] = ((rho_matrices[i].transpose()).times(rho_sqrt)).trace();
            }
            Matrix Y = new Matrix(y,3); // Constructs vector part of matrix equation
            Matrix B = A.solve(Y); // Solves matrix equation
            b_0 = B.get(0,0);
            b_1 = B.get(1,0);
            b_2 = B.get(2,0);
        }
        
        /*
           Prints if the parameters are admissible or not
        */
        
        if (indicator == 1.0) {
            System.out.println("Covariance matrix is positive definite. Paramaters are admissible:");
        } else {
            System.out.println("Covariance matrix is not positive definite. Paramaters are not admissible:");  
        }
        
        /*
           Prints values of parameters
        */
        
        System.out.println("n = " + n);
        System.out.println("a_0 = " + a_0);
        System.out.println("a_1 = " + a_1);
        System.out.println("a_2 = " + a_2);
        
        return new double[]{b_0,b_1,b_2,indicator};
    }
    
    /*
       Sets parameters n, p, t, a_0, a_1, a_2
       Determines if the parameters n, a_0, a_1, a_2 are admissible
    
       If parameters are admissible:
       Uses coefficients of positive definite square root of rho to construct random graph from specified distribution
       Produces Graph object "graph"
       Produces adjacency matrix "adj" as array double[][]
    */
    
    public static void main(String[] args) {
        
        /*
           Parameters
        */
        
        int n = 100; // Number of vertices
        double p = 0.1; // Probability of adding edge
        double t = -1.28; // t is the quantile for p under the standard normal distribution
        double a_0 = 1.0; // Normalization
        double a_1 = 0.1; // p^2*(1+a_1) is covariance of connected motif
        double a_2 = -0.001; // p^2*(1+a_2) is covariance of disjoint motif
        
        /*
           Calls coefficientSqrt
        */
        
        double[] coefficientsqrt = coefficientsSqrt(n,a_0,a_1,a_2);
        double b_0 = coefficientsqrt[0];
        double b_1 = coefficientsqrt[1];
        double b_2 = coefficientsqrt[2];
        double indicator = coefficientsqrt[3];
        
        /*
           If parameters are admissible: Generate random graph as Graph object "graph" and adjacency matrix "adj" as array object double[][]
        */
        
        if (indicator == 1.0) {
            System.out.println("Additional parameters:"); // Print value of p
            System.out.println("p = " + p);
            /*
               Creates symmetric array of i.i.d. standard normal random variables (diagonal remains 0.0)
               Computes sum
            */
            
            double sum = 0;
            Random rand = new Random();
            double[][] random_matrix = new double[n][n];
            for (int i = 0; i < n; i++) {
                for (int j = 0; j < i; j++) {
                    random_matrix[i][j] = rand.nextFloat();
                    random_matrix[j][i] = random_matrix[i][j];
                    sum += random_matrix[i][j];
                }
            }
            
            /*
               Initializes adjacency matrix "adj" as array double[][]
               Initializes Graph object "graph" with n vertices
            */
            
            int[][] adj = new int[n][n];
            Graph graph = new DefaultUndirectedGraph(DefaultEdge.class);
            for (int i = 0; i < n; i++) {
                graph.addVertex(i);
            }
            
            /*
               Computes the contribution of random variables to the edge connecting vertices i and j and compares to threshold t
            */
            
            for (int i = 0; i < n; i++) {
                for(int j = 0; j < i; j++) {
                    double sum_conn = 0;
                    for (int k = 0; k < n; k++) {
                        if ((k != i) && (k != j)) {
                            sum_conn += random_matrix[i][k];
                            sum_conn += random_matrix[j][k];
                        }
                    }
                    if ((b_2 - b_0)*random_matrix[1][0] + (b_1 - b_2)*sum_conn + b_2*sum < t) {
                        adj[i][j] = 1; // Includes edge in adjacency matrix
                        adj[j][i] = 1; // Adjacency matrix is symmetric since graph is undirected
                        graph.addEdge(i, j); // Included edge in graph
                    }
                }
            }
            
            /*
               Final result:
               Graph object "graph" (undirected graph) and double[][] object "adj" (adjacency matrix).
               Can use package jgrapht to display or apply graph algorithms to Graph object "graph".
            */
        }
    }   
}
